src/coreclr/jit/earlyprop.cpp

```diff
@@ -645,15 +648,55 @@ void Compiler::optFoldNullCheck(GenTree* tree, LocalNumberToNullCheckTreeMap* nu
 //       or
 //       indir(add(x, const2))
 //
-//       (indir is any node for which OperIsIndirOrArrLength() is true.)
+//       (indir is any node for which OperIsIndirOrArrLength() or OperIsImplicitIndir() is true.)
 //
 //     2.  const1 + const2 if sufficiently small.
-
+//
 GenTree* Compiler::optFindNullCheckToFold(GenTree* tree, LocalNumberToNullCheckTreeMap* nullCheckMap)
 {
-    assert(tree->OperIsIndirOrArrLength());
+    assert(tree->OperIsDereference());
+
+    GenTree* addr;
+    if (tree->OperIsIndir())
+    {
+        addr = tree->AsIndir()->Addr();
+    }
+    else
+    {
+        switch (tree->OperGet())
+        {
+            case GT_ARR_LENGTH:
+                addr = tree->AsArrLen()->ArrRef();
+                break;
+
+            case GT_XORR:
+            case GT_XAND:
+            case GT_XADD:
+            case GT_XCHG:
+            case GT_LOCKADD:
+            case GT_ARR_INDEX:
+            {
+                const bool commaOnly = true;
+
+                addr = tree->AsOp()->gtGetOp1()->gtEffectiveVal(commaOnly);
+                break;
+            }
+
+            case GT_CMPXCHG:
+                addr = tree->AsCmpXchg()->gtOpLocation;
+                break;
+
+            case GT_BOX:
+                addr = tree->AsBox()->BoxOp();
```

Comment:
so should I delete this check then or return nullptr?
Like 
```
// GT_BOX does not represent an indirection after `impImportAndPushBox`.
return nullptr;
```

Sounds like `OperIsImplicitIndir` also should not return `true` for it after this method.

Also, don't you think that having both `OperIsDereference` and `OperIsIndir` with different meanings is confusing?

Reply:
With our current implementation `GT_BOX` never represents an indirection. It's always just a marker node so it shouldn't be part of `OperIsImplicitIndir`.

It's already confusing, e.g., `GT_BLK`, `GT_OBJ`, `GT_DYN_BLK`, `GT_STORE_BLK`, `GT_STORE_OBJ`, and `GT_STORE_DYN_BLK` are included in both `OperIsIndir` and `OperIsImplicitIndir` while `GT_INDIR` isn't. Also,  `GT_ARR_INDEX`, `GT_ARR_ELEM`, and `GT_ARR_OFFSET` are included in `OperIsImplicitIndir` but `GT_ARR_LENGTH` is not. It would be good to clean this up and make this methods more normalized. We are probably missing some cases in optimizations. I think we should have a helper that includes all of these indirections, not sure what the best name for it is.